Objects/StrategyCardAutomator.ttslua

```diff
@@ -1637,108 +1648,172 @@ function researchTech(color, cardName)
 		return
 	end
 
-	-- searches through container for an item with name matching namePattern and returns its guid
-	local function getEntryGuid(container, namePattern)
-		assert(type(container) == 'userdata' and type(namePattern) == 'string')
-		for _, entry in ipairs(container.getObjects()) do
-			namePattern = string.gsub(namePattern, "-",".")
-			if string.match(entry.name, namePattern) then
-				return entry.guid
-			end
-		end
-		return false
-	end
-
-
 	local factionSheet = getObjectFromGUID(faction.factionSheetGuid)
 	if not factionSheet then
 		broadcastToAll('research Technology: missing ' .. color .. ' Faction Sheet', 'Red')
 		return
 	end
 
-	local overrides = _unitHelper.getUnitOverrides()
-	local pos
-	local rot = {}
-	rot = techDeck.getRotation()
-
-	-- unit upgades are sent to factionSheet, other tech just put left of deck.
-	if overrides[cardName] and overrides[cardName]["upgrade"] then
-		-- code copied from setupHelper factionsheetsnappoints
-		local x0 = 1.09
-		local dx = (-1.09 * 2) / 3
-		local z0 = -0.74
-		local dz = (0.74 * 2) / 3
-
-		local counter = 1
-		local snapPoints = {}
-		for col = 0, 3 do
-			for row = 0, 3 do
-				if col <= row then
-					snapPoints[counter] = {
-							x = x0 + col * dx,
-							y = 0,
-							z = z0 + row * dz
-						}
-					counter = counter + 1
+	local guid = _getEntryGuid(techDeck, cardName)
+	if not guid then
+		_safeBroadcastToColor('Research Technology: ' .. cardName .. " not found in deck " .. techDeck.getName(), color, color)
+		return
+	end
+
+
+
+	local function _getPosToTechBoard(cardName, techBoard, techDeck)
+		local pos
+		local dx
+		local dz
+		-- find grid size and origin of snap points on tech board
+		for _,snap in ipairs(techBoard.getSnapPoints()) do
+			if not pos then
+				pos = snap.position
+			else
+				if (not dz) and (snap.position.z < (pos.z - 0.1)) then
+					dz = snap.position.z - pos.z
+				end
+				if (not dx) and (snap.position.x < (pos.x - 0.1)) then
+					dx = pos.x - snap.position.x
+				end
+				if snap.position.x > pos.x then
+					pos.x = snap.position.x
+				end
+				if snap.position.z > pos.z then
+					pos.z = snap.position.z
 				end
 			end
 		end
 
-		local pointorder = {
-		["Flagship"] = 1,
-		["War Sun"] = 2,
-		["Dreadnought"] = 3,
-		["Carrier"] = 4,
-		["Cruiser"] = 5,
-		["Destroyer"] = 6,
-		["Fighter"] = 7,
-		["PDS"] = 8,
-		["Infantry"] = 9,
-		["Space Dock"] = 10
-		}
+		-- find column for tech color
+		local _columns = {["Blue"]=0, ["Green"]=1, ["Yellow"]=2, ["Red"]=3, ["unitUpgrade"]=4}
+		column = _columns[_technologyHelper.getTechType(cardName)]
+		pos.x = pos.x - column*dx
+
+		-- find first spot not occupied by tech card
```

Comment:
This loop structure could probably be more efficient, e.g. build the set of taken points in the column in a single pass, then find the first available from that smaller set.  This is probably ok as you've already restricted to just tech cards in this player's area.